constraints-oapi.yml

```diff
@@ -0,0 +1,485 @@
+                          "0x02f8908330182402845acf6267845acf75718254ae94614561d2d143621e126e87831aef287678b442b8830186a0a09498dae671e12a4df22225c07db38b1b6c4bf4207b9e49fb10062918fb9d4d8fc001a0a8038bf97cc0f7bc742b7b319b4d98154300eb941746f53c48362c55df1913d6a0728550ffc7c06da44f2d6d356285bc24236eac6628a879d5072321eb73d071e4",
+                          "0x02f86e83301824068082075a825208948943545177806ed17b9f23f0a21ee5948ecaa77687017c6953f31cb880c001a0462fb6848c16e585b786e4fc168bd20a3cb180ebf00e34b93fed9c42b0df0e76a03ac159e02cee545d3e24be5fbd26f902626e18bcf6626f50a7636d4de28834e7"
+                        ]
+                        withdrawals: []
+                        blob_gas_used: "0"
+                        excess_blob_gas: "0"
+                      signature: "0x1b66ac1fb663c9bc59509846d6ec05345bd908eda73e670af888da41af171505cc411d61252fb6cb3fa0017b679f8bb2305b26a285fa2737f175668d0dff91cc1b66ac1fb663c9bc59509846d6ec05345bd908eda73e670af888da41af171505"
+                      proofs:
+                        commitmentTypes: [1, 2]
+                        payloads: [
+                          "0x123456789",  # Example proof payload for commitment type 1
+                          "0x987654321fedcba0"   # Example proof payload for commitment type 2
+                        ]
+                      blobs_bundle:
+                        commitments: []
+                        proofs: []
+                        blobs: []
+      responses:
+        "200":
+          description: "OK"
+          content: {}
+components:
+  schemas:
+    SignedDelegation:
+      type: object
+      properties:
+        message:
+          $ref: '#/components/schemas/Delegation'
+        signature:
+            $ref: "../beacon-apis/types/primitive.yaml#/Signature"
+      required:
+        - message
+        - signature
+    Delegation:
+      type: object
+      properties:
+        proposer:
+            $ref: "../beacon-apis/types/primitive.yaml#/Pubkey"
+        delegate:
+            $ref: "../beacon-apis/types/primitive.yaml#/Pubkey"
+        slasher:
+          type: string
+          description: Ethereum address
+        valid_until:
+          type: integer
+          description: Slot
+        metadata:
+          type: string
+          description: Arbitrary bytes
+      required:
+        - proposer
+        - delegate
+        - slasher
+        - valid_until
+        - metadata
+    SignedConstraints:
+      type: object
+      properties:
+        message:
+          $ref: '#/components/schemas/ConstraintsMessage'
+        signature:
+            $ref: "../beacon-apis/types/primitive.yaml#/Signature"
+      required:
+        - message
+        - signature
+    ConstraintsMessage:
```

Comment:
I suggest we add a `proposer` field to the `ConstraintsMessage`. 

only using `delegate` is insufficient unless the assumption is that the Relay looks up the `proposer` pubkey for that slot number from the beacon api, then checks for a `SignedDelegations` that delegates from the validator to the delegate. This isn’t a terrible assumption but means the API can’t support proposers that aren’t validators (e.g., for building L2 blocks). 

We can either:
a) rename `delegate` to `proposer`. Then the Relay has to verify the `ConstraintsMessage.signature` against **all**  of the `delegate` pubkeys it has received `SignedDelegations` for. This can be done by iterating through `SignedDelegations` from the GET /delegations endpoint but can be expensive.

b) include both `delegate` and `proposer` fields. Then the Relay can lookup `SignedDelegations` by `delegate` pubkeys. 

I am partial to **b)** to minimize Relay complexity and allow the API to support proposers that aren't validators (e.g., for delegating L2 block building rights to issue L2 preconfs without based sequencing).

Reply:
> I am partial to **b)** to minimize Relay complexity and allow the API to support proposers that aren't validators (e.g., for delegating L2 block building rights to issue L2 preconfs without based sequencing).

I agree. I think that L2 interoperability for preconfirmations is an attractive design goal but even without this as a factor, adding a single field to the API types seems much simpler than (indirectly) forcing relays to perform additional interactions via the Beacon API than they would otherwise have to. It's certainly not as if the size of messages across the network is a marginal constraint here anyway.

On a related note,

> his can be done by iterating through `SignedDelegations` from the GET /delegations endpoint but can be expensive.

What do we expect the typical sizes of this set to be in a production setting?